torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -922,15 +905,24 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
     VLOG(2) << "[Rank " << rank_
             << "] NCCL watchdog thread terminated normally";
   } catch (std::exception& e) {
-    // Append error message reported from workCleanupLoop
-    const auto exitMsg = c10::str(
-        "[Rank ",
-        rank_,
-        "] NCCL watchdog thread terminated with exception: ",
-        e.what());
-    LOG(ERROR) << exitMsg;
-    watchDogException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
-    std::rethrow_exception(watchDogException_);
+    if (std::string(e.what()).find("driver shutting down") !=
+        std::string::npos) {
+      LOG(INFO)
+          << "[Rank " << rank_
+          << "] main process destroyed cuda before watchdog loop exited, terminating watchdog."
+          << " (Watchdog caught exception: " << e.what();
+
+    } else {
+      // Append error message reported from workCleanupLoop
+      const auto exitMsg = c10::str(
+          "[Rank ",
+          rank_,
+          "] NCCL watchdog thread terminated with exception: ",
+          e.what());
+      LOG(ERROR) << exitMsg;
+      watchDogException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
+      std::rethrow_exception(watchDogException_);
```

Comment:
I thought that `std::make_exception_ptr` is only needed/applicable if one wants to avoid downcasting type errors. I.e. what's wrong with good old throw?
Also, are there any docs on the lifetime of exception_ptr object? (I.e. I would not expect it to be a valid pointer outside of current stack unroll chain)
```suggestion
      throw watchDogException_ = std::runtime_error(exitMsg);
```

Reply:
tagging @rohan-varma @H-Huang @kwen2501 for any context about this.  (we can probably address it in a later PR but it would be good to follow up)

Reply:
I don't understand that re-throw logic as well... Since this is a refactor, I think it's ok that we do it for now. We can probably add a TODO and I can follow up with the team later to see what is the best way to throw exceptions.